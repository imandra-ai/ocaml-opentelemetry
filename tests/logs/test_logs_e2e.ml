module Client = Opentelemetry_client
module L = Opentelemetry_proto.Logs

(* NOTE: This port must be different from that used by other integration tests,
   to prevent socket binding clashes. *)
let port = 4359

let url = Printf.sprintf "http://localhost:%d" port

let cmd = [ "emit_logs_cohttp"; "--url"; url ]

let tests (signal_batches : Client.Signal.t list) =
  ignore signal_batches;
  List.iter
    (fun (signal_batch : Client.Signal.t) ->
      match signal_batch with
      | Logs ls ->
        ls (* Mask out the times so tests don't change in between runs *)
        |> List.map (fun (l : L.resource_logs) ->
               let masked_scope_logs =
                 List.map
                   (fun (sl : L.scope_logs) ->
                     let masked_log_records =
                       List.map
                         (fun (lr : L.log_record) ->
                           {
                             lr with
                             time_unix_nano = 0L;
                             observed_time_unix_nano = 0L;
                           })
                         sl.log_records
                     in
                     { sl with log_records = masked_log_records })
                   l.scope_logs
               in
               { l with scope_logs = masked_scope_logs })
        |> List.iter (Format.printf "%a\n" L.pp_resource_logs)
      | _ -> ())
    signal_batches

let () =
  let signal_batches = Signal_gatherer.gather_signals ~port cmd in
  tests signal_batches
