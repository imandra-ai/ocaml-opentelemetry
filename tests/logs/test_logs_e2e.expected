{ resource =
    Some(
      { attributes =
          [{ key = "service.namespace";
             value = Some(String_value("ocaml-otel.test"));
           };
           { key = "service.name"; value = Some(String_value("emit_logs")); };
           { key = "src"; value = Some(String_value("application")); };
           { key = "my_reporter_attr"; value = Some(String_value("foo")); }
           ];
        dropped_attributes_count = 0;
      });
  scope_logs =
    [{ scope =
         Some(
           { name = "ocaml-otel";
             version = "%%VERSION_NUM%%";
             attributes = [];
             dropped_attributes_count = 0;
           });
       log_records =
         [{ time_unix_nano = 0;
            observed_time_unix_nano = 0;
            severity_number = Severity_number_debug;
            severity_text = "debug";
            body = Some(String_value("emit_logs: starting"));
            attributes = [];
            dropped_attributes_count = 0;
            flags = 0;
            trace_id = <bytes len=0>;
            span_id = <bytes len=0>;
          }
          ];
       schema_url = "";
     }
     ];
  schema_url = "";
}
{ resource =
      Some(
        { attributes =
            [{ key = "service.namespace";
               value = Some(String_value("ocaml-otel.test"));
             };
             { key = "service.name"; value = Some(String_value("emit_logs")); };
             { key = "src"; value = Some(String_value("application")); };
             { key = "my_reporter_attr"; value = Some(String_value("foo")); }
             ];
          dropped_attributes_count = 0;
        });
    scope_logs =
      [{ scope =
           Some(
             { name = "ocaml-otel";
               version = "%%VERSION_NUM%%";
               attributes = [];
               dropped_attributes_count = 0;
             });
         log_records =
           [{ time_unix_nano = 0;
              observed_time_unix_nano = 0;
              severity_number = Severity_number_info2;
              severity_text = "info";
              body = Some(String_value("emit_logs: info log"));
              attributes = [];
              dropped_attributes_count = 0;
              flags = 0;
              trace_id = <bytes len=0>;
              span_id = <bytes len=0>;
            }
            ];
         schema_url = "";
       }
       ];
    schema_url = "";
  }
{ resource =
        Some(
          { attributes =
              [{ key = "service.namespace";
                 value = Some(String_value("ocaml-otel.test"));
               };
               { key = "service.name";
                 value = Some(String_value("emit_logs"));
               };
               { key = "src"; value = Some(String_value("application")); };
               { key = "my_reporter_attr"; value = Some(String_value("foo")); }
               ];
            dropped_attributes_count = 0;
          });
      scope_logs =
        [{ scope =
             Some(
               { name = "ocaml-otel";
                 version = "%%VERSION_NUM%%";
                 attributes = [];
                 dropped_attributes_count = 0;
               });
           log_records =
             [{ time_unix_nano = 0;
                observed_time_unix_nano = 0;
                severity_number = Severity_number_warn;
                severity_text = "warning";
                body = Some(String_value("emit_logs: warn log"));
                attributes = [];
                dropped_attributes_count = 0;
                flags = 0;
                trace_id = <bytes len=0>;
                span_id = <bytes len=0>;
              }
              ];
           schema_url = "";
         }
         ];
      schema_url = "";
    }
{ resource =
          Some(
            { attributes =
                [{ key = "service.namespace";
                   value = Some(String_value("ocaml-otel.test"));
                 };
                 { key = "service.name";
                   value = Some(String_value("emit_logs"));
                 };
                 { key = "src"; value = Some(String_value("application")); };
                 { key = "my_reporter_attr";
                   value = Some(String_value("foo"));
                 }
                 ];
              dropped_attributes_count = 0;
            });
        scope_logs =
          [{ scope =
               Some(
                 { name = "ocaml-otel";
                   version = "%%VERSION_NUM%%";
                   attributes = [];
                   dropped_attributes_count = 0;
                 });
             log_records =
               [{ time_unix_nano = 0;
                  observed_time_unix_nano = 0;
                  severity_number = Severity_number_error;
                  severity_text = "error";
                  body = Some(String_value("emit_logs: error log"));
                  attributes = [];
                  dropped_attributes_count = 0;
                  flags = 0;
                  trace_id = <bytes len=0>;
                  span_id = <bytes len=0>;
                }
                ];
             schema_url = "";
           }
           ];
        schema_url = "";
      }
{ resource =
            Some(
              { attributes =
                  [{ key = "service.namespace";
                     value = Some(String_value("ocaml-otel.test"));
                   };
                   { key = "service.name";
                     value = Some(String_value("emit_logs"));
                   };
                   { key = "src"; value = Some(String_value("application")); };
                   { key = "my_reporter_attr";
                     value = Some(String_value("foo"));
                   }
                   ];
                dropped_attributes_count = 0;
              });
          scope_logs =
            [{ scope =
                 Some(
                   { name = "ocaml-otel";
                     version = "%%VERSION_NUM%%";
                     attributes = [];
                     dropped_attributes_count = 0;
                   });
               log_records =
                 [{ time_unix_nano = 0;
                    observed_time_unix_nano = 0;
                    severity_number = Severity_number_info;
                    severity_text = "app";
                    body = Some(String_value("emit_logs: app log"));
                    attributes = [];
                    dropped_attributes_count = 0;
                    flags = 0;
                    trace_id = <bytes len=0>;
                    span_id = <bytes len=0>;
                  }
                  ];
               schema_url = "";
             }
             ];
          schema_url = "";
        }
{ resource =
              Some(
                { attributes =
                    [{ key = "service.namespace";
                       value = Some(String_value("ocaml-otel.test"));
                     };
                     { key = "service.name";
                       value = Some(String_value("emit_logs"));
                     };
                     { key = "src";
                       value = Some(String_value("application"));
                     };
                     { key = "string_list_attr";
                       value = Some(String_value(""));
                     };
                     { key = "bool_attr"; value = Some(String_value("")); };
                     { key = "float_attr"; value = Some(String_value("")); };
                     { key = "int_attr"; value = Some(String_value("")); };
                     { key = "string_attr"; value = Some(String_value("")); };
                     { key = "my_reporter_attr";
                       value = Some(String_value("foo"));
                     }
                     ];
                  dropped_attributes_count = 0;
                });
            scope_logs =
              [{ scope =
                   Some(
                     { name = "ocaml-otel";
                       version = "%%VERSION_NUM%%";
                       attributes = [];
                       dropped_attributes_count = 0;
                     });
                 log_records =
                   [{ time_unix_nano = 0;
                      observed_time_unix_nano = 0;
                      severity_number = Severity_number_info2;
                      severity_text = "info";
                      body =
                        Some(
                          String_value(
                            "emit_logs: this log is emitted with varied tags from a span"));
                      attributes = [];
                      dropped_attributes_count = 0;
                      flags = 0;
                      trace_id = <bytes len=16>;
                      span_id = <bytes len=8>;
                    }
                    ];
                 schema_url = "";
               }
               ];
            schema_url = "";
          }
{ resource =
                Some(
                  { attributes =
                      [{ key = "service.namespace";
                         value = Some(String_value("ocaml-otel.test"));
                       };
                       { key = "service.name";
                         value = Some(String_value("emit_logs"));
                       };
                       { key = "src";
                         value = Some(String_value("application"));
                       };
                       { key = "string_list_attr";
                         value = Some(String_value(""));
                       };
                       { key = "bool_attr"; value = Some(String_value("")); };
                       { key = "float_attr"; value = Some(String_value("")); };
                       { key = "int_attr"; value = Some(String_value("")); };
                       { key = "string_attr"; value = Some(String_value("")); };
                       { key = "my_reporter_attr";
                         value = Some(String_value("foo"));
                       }
                       ];
                    dropped_attributes_count = 0;
                  });
              scope_logs =
                [{ scope =
                     Some(
                       { name = "ocaml-otel";
                         version = "%%VERSION_NUM%%";
                         attributes = [];
                         dropped_attributes_count = 0;
                       });
                   log_records =
                     [{ time_unix_nano = 0;
                        observed_time_unix_nano = 0;
                        severity_number = Severity_number_info2;
                        severity_text = "info";
                        body =
                          Some(
                            String_value(
                              "emit_logs: this log will be emitted with varied tags"));
                        attributes = [];
                        dropped_attributes_count = 0;
                        flags = 0;
                        trace_id = <bytes len=0>;
                        span_id = <bytes len=0>;
                      }
                      ];
                   schema_url = "";
                 }
                 ];
              schema_url = "";
            }
{ resource =
                  Some(
                    { attributes =
                        [{ key = "service.namespace";
                           value = Some(String_value("ocaml-otel.test"));
                         };
                         { key = "service.name";
                           value = Some(String_value("emit_logs_fmt"));
                         };
                         { key = "src";
                           value = Some(String_value("application"));
                         };
                         { key = "my_fmt_attr";
                           value = Some(String_value("bar"));
                         }
                         ];
                      dropped_attributes_count = 0;
                    });
                scope_logs =
                  [{ scope =
                       Some(
                         { name = "ocaml-otel";
                           version = "%%VERSION_NUM%%";
                           attributes = [];
                           dropped_attributes_count = 0;
                         });
                     log_records =
                       [{ time_unix_nano = 0;
                          observed_time_unix_nano = 0;
                          severity_number = Severity_number_info2;
                          severity_text = "info";
                          body =
                            Some(
                              String_value(
                                "emit_logs: this log will be emitted from otel and fmt reporter"));
                          attributes = [];
                          dropped_attributes_count = 0;
                          flags = 0;
                          trace_id = <bytes len=0>;
                          span_id = <bytes len=0>;
                        }
                        ];
                     schema_url = "";
                   }
                   ];
                schema_url = "";
              }
